src/charm.py

```diff
@@ -90,45 +113,76 @@ def __init__(self, *args):
         # Expose the Kiali workload through the service
         self.unit.set_ports(Port("tcp", KIALI_PORT))
 
-    def on_collect_status(self, e: ops.CollectStatusEvent):
-        """Collect and report the statuses of the charm."""
-        # TODO: Extract the generation of these statuses to a helper method, that way other parts of the charm can know
-        #  if the charm is active too.
+        # TODO: Once this exists in grafana
+        self._grafana_metadata = GrafanaMetadataRequirer(...)
+
+        # TODO: Once this exists in Tempo
+        self._tempo_metadata = TempoMetadataRequirer(...)
+        self._tempo_datasource_exchange = DatasourceExchange(
+            charm=self,
+            requirer_endpoint=TEMPO_DATASOURCE_EXCHANGE_RELAION_NAME
+        )
+        self.framework.observe(self.on[TEMPO_DATASOURCE_EXCHANGE_RELAION_NAME].relation_changed, self.reconcile)
+
+    def reconcile(self, _event: ops.ConfigChangedEvent):
```

Comment:
The approach used here is to:
* try as much as possible, collecting any statuses that might result in.  So if we can do A but not B, generate a status that reflects why B can't be done
* at the end of reconcile, set the unit status to the worst thing we've encountered and log the rest